File: Permutations/Permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        generatePermutations(ans, list, nums);
        return ans;
    }
    public void generatePermutations(List<List<Integer>> ans, List<Integer> list, int[] nums){
        if (nums.length == 0){
            List<Integer> l = new ArrayList<Integer>(list);
            ans.add(l);
            return;
        }
        for (int i = 0; i < nums.length; i++){
            int [] numsRemaining = new int [nums.length - 1];
            for (int j = 0, k = 0; j < nums.length; j++){
                if (j == i){
                    continue;
                }
                numsRemaining[k++] = nums[j];
            }
            list.add(nums[i]);
            generatePermutations(ans, list, numsRemaining);
            list.remove(list.size() - 1);
        }
        return;
    }
}

File: Coding Problems & My Solutions/Permutations/Permutations.swift
class Solution {
    func permute(_ nums: [Int]) -> [[Int]] {
        var ans: [[Int]] = []
        generate(nums, [], &ans)
        return ans
    }
    func generate(_ nums: [Int], _ perm: [Int], _ ans: inout [[Int]]) {
        if nums.isEmpty {
            ans.append(perm)
        }
        for num in nums {
            var tempNums = nums.filter { $0 != num }
            var tempPerm = perm
            tempPerm.append(num)
            generate(tempNums, tempPerm, &ans)
        }
    }
}